File: in_Java/0560 Subarray Sum Equals K.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        if(nums == null || nums.length == 0) return 0;
        if(nums.length == 1){
            if(nums[0] == k) return 1;
            else return 0;
        }
        int res = 0;
        for(int i = 1; i < nums.length; i++){
            nums[i] = nums[i] + nums[i-1];
        }
        HashMap<Integer,Integer> hm = new HashMap<>();
        for(int i : nums){
            if(hm.containsKey(i - k)) res += hm.get(i-k);
            if(!hm.containsKey(i)) hm.put(i,1);
            else hm.put(i, hm.get(i)+1);
        }
        if(hm.containsKey(k)) res += hm.get(k);
        return res;
    }
}

File: in_Python/0560 Subarray Sum Equals K.py
class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        res = 0
        d = {}
        d[nums[0]] = [0]
        for i in range(1, len(nums)):
            cursum = nums[i-1] + nums[i]
            if cursum not in d:
                t = [i]
                d[cursum] = t
            else:
                t = d[cursum]
                t.append(i)
                d[cursum] = t
            nums[i] = cursum
        #print(d)
        if k in d:
            res += len(d[k])
        if k == 0:
            for key in d:
                n = len(d[key])
                res += n * (n-1) / 2
        else:
            for key in d:
                a = d[key]
                if key + k in d:
                    b = d[key+k]
                    for x in a:
                        for y in b:
                            if y > x:
                                res += 1
        return res